monai/transforms/utility/array.py

```diff
@@ -1715,3 +1715,126 @@ def __call__(self, img: NdarrayOrTensor, meta_dict: Mapping | None = None) -> Nd
         if self._do_transform:
             img = self.filter(img)
         return img
+
+
+class ApplyTransformToPoints(InvertibleTransform, Transform):
+    """
+    Transform points between image coordinates and world coordinates.
+
+    Args:
+        dtype: The desired data type for the output.
+        affine: A 3x3 or 4x4 affine transformation matrix applied to points. Typically, this matrix originates from the image.
+        invert_affine: Whether to invert the affine transformation matrix applied to the points. Defaults to ``True``.
+            Typically, the affine matrix is derived from the image, while the points are in world coordinates.
+            If you want to align the points with the image, set this to ``True``. Otherwise, set it to ``False``.
+        affine_lps_to_ras: Defaults to ``False``. Set this to ``True`` if all of the following are true:
+            1) The image is read by `ITKReader`,
+            2) The `ITKReader` has `affine_lps_to_ras=True`,
+            3) The data is in world coordinates.
+            This ensures the correct application of the affine transformation between LPS (left-posterior-superior)
+            and RAS (right-anterior-superior) coordinate systems. This argument ensures the points and the affine
+            matrix are in the same coordinate system.
+    """
+
+    def __init__(
+        self,
+        dtype: DtypeLike | torch.dtype = torch.float64,
+        affine: torch.Tensor | None = None,
+        invert_affine: bool = True,
+        affine_lps_to_ras: bool = False,
+    ) -> None:
+        self.dtype = dtype
+        self.affine = affine
+        self.invert_affine = invert_affine
+        self.affine_lps_to_ras = affine_lps_to_ras
+
+    @staticmethod
+    def apply_affine_to_points(data: torch.Tensor, affine: torch.Tensor, dtype: DtypeLike | torch.dtype):
+        data_: torch.Tensor = convert_to_tensor(data, track_meta=False, dtype=torch.float64)
+        affine = to_affine_nd(data_.shape[-1], affine)
+
+        homogeneous = concatenate((data_, torch.ones((data_.shape[0], data_.shape[1], 1))), axis=2)
+        transformed_homogeneous = torch.matmul(affine, homogeneous.transpose(1, 2))
+        transformed_coordinates = transformed_homogeneous[:, :-1, :].transpose(1, 2)
+        out, *_ = convert_to_dst_type(transformed_coordinates, data, dtype=dtype)
+
+        return out
```

Comment:
This could be in a utility function instead, there may be need to apply such transforms in other contexts. Also you can avoid doing transpositions of the data by doing matmul with the transposed affine as the second arg, ie. `torch.matmul(homogeneous, affine.T)` or something like this. Also (but check my math) if the bottom row of the matrix is `[0, 0, 0, 1]` you don't need to add the fourth component, just matmul with the upper 3x3 of `affine` then add the translation component of the matrix.

Reply:
Hi @ericspod, agreed and I moved it under transforms/utils.